### DISCUSSION TITLE
Proper count tokens for conversation / message level

### DISCUSSION THREAD

**Discussion by vectorkovacspeter** [danny-avila/LibreChat#1034] [Q&A]
### Contact Details

kovacs.peter@vector.hu

### What is your question?

How can I properly get the input and output token numbers for a conversation or a message?

In the Mongo database, thez messages.tokenCount value is null for the non-user messages, am I search it in the wrong place?

Is there a detailed documentation available for the database schama, and its content?

### More Details

I try to sum the tokens used by a specified conversation or a specified message stored in the chat database (I use mongo).

### What is the main subject of your question?

Other

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by danny-avila** [comment]
Hi @vectorkovacspeter, thanks for trying out LibreChat. 

Are you on the latest commit on main branch? The message.tokenCount is mainly for caching the tokenCount on subsequent message requests.

> the Mongo database, thez messages.tokenCount value is null for the non-user messages, 

This should not be the case, as you can see below

![image](https://github.com/danny-avila/LibreChat/assets/110412045/2c77b850-dbb5-4f79-884e-783c5a4bd392)

> Is there a detailed documentation available for the database schama, and its content?

No, but you can inspect the schemas in [/LibreChat/api/models/schema/](https://github.com/danny-avila/LibreChat/tree/b3aac97710ab9680046eb8089c5fcd4456bd2988/api/models/schema)

To view what is actually spent, you should be looking at the transactions collection and filter by conversationId

**Reply by vectorkovacspeter** [reply]
Thanks for your answer, I use the backend and client version: 0.5.8.

As I checked the database again, not all the tokenCount is null, where the sender is ChatGPT, that some of them. It seems its appears on different models, and strop reasons, so I can't find the common thing in them.

I try to update my version, maybe it's solve the problem, but as I customize it som points, its a bit time consuming to update :)

**Comment by danny-avila** [comment]
See these docs and PR for more info:

https://docs.librechat.ai/features/token_usage.html

https://github.com/danny-avila/LibreChat/pull/1018
